pandas/src/generate_code.py

```diff
@@ -2253,19 +2253,19 @@ def put2d_%(name)s_%(dest_type)s(ndarray[%(c_type)s, ndim=2, cast=True] values,
 def generate_put_template(template, use_ints=True, use_floats=True,
                           use_objects=False, use_datelikes=False):
     floats_list = [
-        ('float64', 'float64_t', 'float64_t', 'np.float64'),
-        ('float32', 'float32_t', 'float32_t', 'np.float32'),
+        ('float64', 'float64_t', 'float64_t', 'np.float64', 'nan'),
+        ('float32', 'float32_t', 'float32_t', 'np.float32', 'nan'),
     ]
     ints_list = [
-        ('int8',  'int8_t',  'float32_t', 'np.float32'),
-        ('int16', 'int16_t', 'float32_t', 'np.float32'),
-        ('int32', 'int32_t', 'float64_t', 'np.float64'),
-        ('int64', 'int64_t', 'float64_t', 'np.float64'),
+        ('int8',  'int8_t',  'int32_t', 'np.int32', 'iNaT'),
```

Comment:
This is going to cause breakage (overflow, rounding errors) if it makes arithmetic operations on integers return integers. Consider `mean([1, 2])`. You can only do this safely for operations like `first`/`last`. Otherwise, casting output to float is the right thing to do.